Moved action buttons to a new bottom toolbar (#4435)

reisfmb opened the pull request.

reisfmb committed:
Moved action buttons to a new bottom toolbar (#4435)

reisfmb commented:
@ashklianko `testInputTypes` have failed because now there's no "close button", only a gray "cancel button" which does not have the class that is used by the test. How to proceed?

ashklianko commented:
> @ashklianko `testInputTypes` have failed because now there's no "close button", only a gray "cancel button" which does not have the class that is used by the test. How to proceed?

@sgauruseu will update test

Commit:
update ui-tests for Image Editor

reisfmb committed:
Fixed bottom toolbar edit mode behavior (#4435)

ashklianko commented on modules/lib/src/main/resources/assets/js/app/inputtype/ui/selector/image/ImageEditor.ts:
```diff
@@ -788,6 +751,11 @@ export class ImageEditor
         this.uploadButton.setEnabled(false).addClass('button-upload');
         standbyContainer.appendChildren(this.resetButton, this.uploadButton);
 
+        let rightContainer = new DivEl('right-container');
```
1. Don't hesitate using const, thus we declare that variable is not going to change
2. Always set type of a variable, not only in new code, but also when you update some piece of code where vars are missing their types or have `let` instead of `const`

reisfmb replied:
My bad here. I extracted the code from the top toolbar to create the bottom one, so I wasn't paying attention to those details. Thanks!

ashklianko commented on modules/lib/src/main/resources/assets/js/app/inputtype/ui/selector/image/ImageEditor.ts:
```diff
@@ -788,6 +751,11 @@ export class ImageEditor
+        let rightContainer = new DivEl('right-container');
+        rightContainer.appendChildren(standbyContainer/* , editContainer */);
+
+        let zoomContainer = this.createZoomContainer();
```
Same as above

ashklianko commented on modules/lib/src/main/resources/assets/js/app/inputtype/ui/selector/image/ImageEditor.ts:
```diff
@@ -856,6 +819,53 @@ export class ImageEditor
     }
 
+    private createBottomToolbar(): DivEl {
+        let toolbar = new DivEl('bottom-toolbar');
```
Same: `const` instead of `let`, and variable types

ashklianko commented on modules/lib/src/main/resources/assets/js/app/inputtype/ui/selector/image/ImageEditor.ts:
```diff
@@ -856,6 +819,53 @@ export class ImageEditor
+    private createBottomToolbar(): DivEl {
+        let toolbar = new DivEl('bottom-toolbar');
+
+        let editContainer = new DivEl('edit-container');
```
Same as above

ashklianko commented on modules/lib/src/main/resources/assets/js/app/inputtype/ui/selector/image/ImageEditor.ts:
```diff
@@ -856,6 +819,53 @@ export class ImageEditor
+            }
+        });
+
+        let applyButton = new Button(i18n('button.apply'));
```
Same as above

ashklianko commented on modules/lib/src/main/resources/assets/js/app/inputtype/ui/selector/image/ImageEditor.ts:
```diff
@@ -856,6 +819,53 @@ export class ImageEditor
+            }
+        });
+
+        let cancelButton = new Button(i18n('action.cancel'));
```
Same as above

ashklianko commented:
One more thing: lets replace crop drag handler (circle) that is might not be seen on the image depending on it's color with the next drag control on the bottom toolbar:

![image](https://user-images.githubusercontent.com/4255706/163267806-6285f81d-5b28-4910-803f-e0483552e2f2.png)

Use class `icon-arrow-top` on the element representing button (div or whatever), with `@admin-medium-gray-border` color

reisfmb committed:
Small code adjustments, let to const and typing (#4435)

reisfmb committed:
Improved drag handler (#4435)

ashklianko committed:
Moved action buttons to a new bottom toolbar (#4435)

Commit:
update failed ui-tests for embed macro

reisfmb commented:
Remains to update position of the drag handler in crop mode. It should be inside the bottom toolbar. Will do that as 1st thing on Monday.

reisfmb committed:
Adjusted drag handler placement (#4435)

+ Made an adjustment to the drag event

+ Handled the case when drag handler would overlap buttons in bottom toolbar

alansemenov reviewed (changes requested):
Not exactly right. 
1. The drag control should be a circle with two arrows facing up and down. See below for how it was before.
2. When you drag the control up/down, the footer should stay put - it's only the line with the control that should change position.
<img width="1674" alt="Screenshot 2022-04-18 at 23 56 28" src="https://user-images.githubusercontent.com/11612490/163884433-44f79a24-b4ab-4249-b7a3-d8ec01accfa1.png">

ashklianko commented:
@alansemenov we've tried with keeping bottom toolbar at place, that didn't look good, also drag handler may blend into the background

reisfmb commented:
@alansemenov just to add more context...

1. Opted to go with Sasha's suggestion since the drag handler might blend with the image if it isn't fully in the bottom toolbar. 

2. Last week I recorded a video and asked Sasha's opinion on it... we both agreed that it didn't look good to have the bottom toolbar at place.

https://user-images.githubusercontent.com/67838246/164074061-fb94fcdf-24ee-44d3-a3d6-f9bd607e31df.mp4

How to proceed?

alansemenov commented:
Ok, I agree about the moving bottom toolbar, but dragger has to be changed. The argument that it might blend with the image doesn't stand when you look at the 50 shades of grey in the image below. I had a really hard time realising the dragger was even there. Let's make it the way it was (see a screenshot above) and invert the colors inside, so that the circle is dark and arrows inside are the same grey color as the toolbar.

![image](https://user-images.githubusercontent.com/11612490/164461786-5a54c94d-d348-4cd1-b656-1443991d4001.png)

alansemenov commented:
Made a new issue: https://github.com/enonic/app-contentstudio/issues/4557
